docs/book/kafka-authorization.md

```diff
@@ -0,0 +1,332 @@
+
+Finally, run `kafka-console-consumer` again but this time try to use the
+`anon_consumer` credentials. The `anon_consumer` credentials simulate a
+service that has **not** been explicitly granted access to PII data.
+
+```bash
+docker run --rm --network opakafkatutorial_default \
+    openpolicyagent/demo-kafka:1.0 \
+    kafka-console-consumer --bootstrap-server kafka:9093 --topic credit-scores --from-beginning --consumer.config /etc/kafka/secrets/anon_consumer.ssl.config
+```
+
+Because the `anon_consumer` is not allowed to read PII data, the request will
+be denied and the consumer will output an error message.
+
+```
+Not authorized to read from topic credit-scores.
+...
+Processed a total of 0 messages
+```
+
+### 4. Extend the policy to prevent services from accidentally writing to topics with large fanout.
+
+First, add the following content to the policy file (`./policies/tutorial.rego`):
+
+
+```ruby
+deny {
+  is_write_operation
+  topic_has_large_fanout
+  not producer_is_whitelisted_for_large_fanout
+}
+
+producer_whitelist = {
+  "large-fanout": {
+    "fanout_producer",
+  }
+}
+
+topic_has_large_fanout {
+  topic_metadata[topic_name].tags[_] == "large-fanout"
+}
+
+producer_is_whitelisted_for_large_fanout {
+  producer_whitelist["large-fanout"][_] == principal.name
+}
+```
+
+Next, update the `topic_metadata` data structure in the same file to indicate
+that the `click-stream` topic has a high fanout.
+
+```ruby
+topic_metadata = {
+  "click-stream": {
+    "tags": ["large-fanout"],
+  },
+  "credit-scores": {
+    "tags": ["pii"],
+  }
+}
+```
+
+### 5. Exercise the policy that restricts producer access to topics with high fanout.
+
+First, run `kafka-console-producer` and simulate a service with access to the
+`click-stream` topic.
+
+```bash
+docker run --rm --network opakafkatutorial_default \
+    openpolicyagent/demo-kafka:1.0 \
+    bash -c 'for i in {1..10}; do echo "{\"user\": \"alice\", \"button\": $i}"; done | kafka-console-producer --topic click-stream --broker-list kafka:9093 -producer.config /etc/kafka/secrets/fanout_producer.ssl.config'
+```
+
+Next, run the `kafka-console-consumer` to confirm that the messages were published.
+
+```bash
+docker run --rm --network opakafkatutorial_default \
+    openpolicyagent/demo-kafka:1.0 \
+    kafka-console-consumer --bootstrap-server kafka:9093 --topic click-stream --from-beginning --consumer.config /etc/kafka/secrets/anon_consumer.ssl.config
+```
+
+Once you see the 10 messages produced by the first part of this step, exit the console consumer (^C).
+
+Lastly, run `kafka-console-producer` to simulate a service that should **not**
+have access to _high fanout_ topics.
+
+```bash
+docker run --rm --network opakafkatutorial_default \
+    openpolicyagent/demo-kafka:1.0 \
+    bash -c 'echo "{\"user\": \"alice\", \"button\": \"bogus\"}" | kafka-console-producer --topic click-stream --broker-list kafka:9093 -producer.config /etc/kafka/secrets/anon_producer.ssl.config'
+```
+
```

Comment:
Can we add some part of the output of this command:
```
[2018-04-26 07:56:38,264] ERROR Error when sending message to topic click-stream with key: null, value: 36 bytes with error: (org.apache.kafka.clients.producer.internals.ErrorLoggingCallback)
org.apache.kafka.common.errors.TopicAuthorizationException: Not authorized to access topics: [click-stream]```

Reply:
Done.